File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/search/SearchResponse.java
package fr.myprysm.vertx.elasticsearch.action.search;

import fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation;
import fr.myprysm.vertx.elasticsearch.action.search.suggest.Suggestion;
import fr.myprysm.vertx.elasticsearch.action.support.ShardFailure;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.util.List;
import java.util.Map;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject
public class SearchResponse {

    private int totalShards;
    private int successfulShards;
    private int skippedShards;
    private int failedShards;
    private List<ShardFailure> shardFailures;
    private Clusters clusters;
    private long tookInMillis;
    private String scrollId;
    private SearchHits hits;
    private Map<String, Aggregation> aggregations;
    private Map<String, Suggestion> suggest;
    private Map<String, JsonObject> profileResults;
    private boolean timedOut;
    private Boolean terminatedEarly;
    private int numReducePhases;

    public SearchResponse(SearchResponse other) {
        totalShards = other.totalShards;
        successfulShards = other.successfulShards;
        skippedShards = other.skippedShards;
        shardFailures = other.shardFailures;
        clusters = other.clusters;
        tookInMillis = other.tookInMillis;
        scrollId = other.scrollId;
        hits = other.hits;
        aggregations = other.aggregations;
        suggest = other.suggest;
        profileResults = other.profileResults;
        timedOut = other.timedOut;
        terminatedEarly = other.terminatedEarly;
        numReducePhases = other.numReducePhases;
    }

    @SuppressWarnings("unchecked")
    public <T extends Aggregation> T getAggregationByName(String name) {
        if (aggregations == null) {
            return null;
        }

        return (T) aggregations.get(name);

    }

    @SuppressWarnings("unchecked")
    public <T extends Suggestion> T getSuggestionByName(String name) {
        if (suggest == null) {
            return null;
        }

        return (T) suggest.get(name);
    }

    /**
     * Build a new <code>SearchResponse</code> from a <code>JsonObject</code>.
     *
     * @param json the <code>JsonObject</code>
     */
    public SearchResponse(JsonObject json) {
        SearchResponseConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>SearchResponse</code> into a <code>JsonObject</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        SearchResponseConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/search/SearchResponse.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.search

import fr.myprysm.vertx.elasticsearch.action.search.SearchResponse
import fr.myprysm.vertx.elasticsearch.action.search.Clusters
import fr.myprysm.vertx.elasticsearch.action.search.SearchHits
import fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation
import fr.myprysm.vertx.elasticsearch.action.search.suggest.Suggestion
import fr.myprysm.vertx.elasticsearch.action.support.ShardFailure

fun SearchResponse(
  aggregations: Map<String, fr.myprysm.vertx.elasticsearch.action.search.aggregations.Aggregation>? = null,
  clusters: fr.myprysm.vertx.elasticsearch.action.search.Clusters? = null,
  failedShards: Int? = null,
  hits: fr.myprysm.vertx.elasticsearch.action.search.SearchHits? = null,
  numReducePhases: Int? = null,
  profileResults: Map<String, io.vertx.core.json.JsonObject>? = null,
  scrollId: String? = null,
  shardFailures: Iterable<fr.myprysm.vertx.elasticsearch.action.support.ShardFailure>? = null,
  skippedShards: Int? = null,
  successfulShards: Int? = null,
  suggest: Map<String, fr.myprysm.vertx.elasticsearch.action.search.suggest.Suggestion>? = null,
  terminatedEarly: Boolean? = null,
  timedOut: Boolean? = null,
  tookInMillis: Long? = null,
  totalShards: Int? = null): SearchResponse = fr.myprysm.vertx.elasticsearch.action.search.SearchResponse().apply {

  if (aggregations != null) {
    this.setAggregations(aggregations)
  }
  if (clusters != null) {
    this.setClusters(clusters)
  }
  if (failedShards != null) {
    this.setFailedShards(failedShards)
  }
  if (hits != null) {
    this.setHits(hits)
  }
  if (numReducePhases != null) {
    this.setNumReducePhases(numReducePhases)
  }
  if (profileResults != null) {
    this.setProfileResults(profileResults)
  }
  if (scrollId != null) {
    this.setScrollId(scrollId)
  }
  if (shardFailures != null) {
    this.setShardFailures(shardFailures.toList())
  }
  if (skippedShards != null) {
    this.setSkippedShards(skippedShards)
  }
  if (successfulShards != null) {
    this.setSuccessfulShards(successfulShards)
  }
  if (suggest != null) {
    this.setSuggest(suggest)
  }
  if (terminatedEarly != null) {
    this.setTerminatedEarly(terminatedEarly)
  }
  if (timedOut != null) {
    this.setTimedOut(timedOut)
  }
  if (tookInMillis != null) {
    this.setTookInMillis(tookInMillis)
  }
  if (totalShards != null) {
    this.setTotalShards(totalShards)
  }
}

